### PR TITLE
feat(admission): validate regex supplied in HTTPRoute (stop rejecting)

### PR DESCRIPTION
**Pull request by programmer04**
**What this PR does / why we need it**:

It introduces better validation for `HTTPRoute` object:
- uses the Kong Gateway endpoint for validating `KongRoute` by calling method [RouteService.Validate](https://pkg.go.dev/github.com/kong/go-kong/kong#RouteService.Validate)
- stops rejecting
     - regex path matching rules (implemented https://github.com/Kong/kubernetes-ingress-controller/issues/2153)
     - regex header matching rules (implemented https://github.com/Kong/kubernetes-ingress-controller/issues/2154)
- still reject - query parameters matching (not implemented yet https://github.com/Kong/kubernetes-ingress-controller/issues/3679)
- introduces test coverage for both  expressions based routers and old ones

**Which issue this PR fixes**:

This is part of https://github.com/Kong/kubernetes-ingress-controller/issues/4557

**Special notes for your reviewer**:

Since integration tests are run in a matrix for expressions router and old ones it can be covered with tests that are exclusively run when the particular router is used. As it's been written in the comment, for non-expression router regex expressions for headers are not validated (require further investigation).

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: test/integration/httproute_webhook_test.go [hunk 1]
```diff
@@ -139,13 +88,197 @@ func TestHTTPRouteValidationWebhook(t *testing.T) {
...
+	testHTTPRouteValidationWebhook(ctx, t, namespace, gatewayClient, testCases)
+}
+
+func setUpEnvForTestingHTTPRouteValidationWebhook(ctx context.Context, t *testing.T) (
+	namespace string,
+	gatewayClient *gatewayclient.Clientset,
+	managedGateway *gatewayv1beta1.Gateway,
+	unmanagedGateway *gatewayv1beta1.Gateway,
+) {
+	closer, err := ensureAdmissionRegistration(
```
Comment (pmalek) [inline test/integration/httproute_webhook_test.go:219]: We could add a `t *testing.T` param to `ensureAdmissionRegistration` and use `t.Cleanup` instead of a `closer`

File: test/integration/httproute_webhook_test.go [hunk 2]
```diff
@@ -139,13 +88,197 @@ func TestHTTPRouteValidationWebhook(t *testing.T) {
...
+		assert.NoError(t, closer())
+	})
+
+	t.Log("creating a gateway client")
+	gatewayClient, err = gatewayclient.NewForConfig(env.Cluster().Config())
+	require.NoError(t, err)
+
+	ns, cleaner := helpers.Setup(ctx, t, env)
+	namespace = ns.Name
+	t.Log("creating a managed gateway")
```
Comment (pmalek) [inline test/integration/httproute_webhook_test.go:244]: Might as well log the gateway name after successful creation.

File: test/integration/httproute_webhook_test.go [hunk 3]
```diff
@@ -139,13 +88,197 @@ func TestHTTPRouteValidationWebhook(t *testing.T) {
...
+
+	t.Log("creating an unmanaged gatewayclass")
+	unmanagedGatewayClass, err := DeployGatewayClass(ctx, gatewayClient, uuid.NewString(), func(gc *gatewayv1beta1.GatewayClass) {
+		gc.Spec.ControllerName = unsupportedControllerName
+	})
+	require.NoError(t, err)
+	cleaner.Add(unmanagedGatewayClass)
+
+	t.Log("creating an unmanaged gateway")
+	unmanagedGateway, err = DeployGateway(ctx, gatewayClient, ns.Name, unmanagedGatewayClass.Name, func(g *gatewayv1beta1.Gateway) {
```
Comment (pmalek) [inline test/integration/httproute_webhook_test.go:259]: ditto

File: test/integration/httproute_webhook_test.go [hunk 4]
```diff
@@ -139,13 +88,197 @@ func TestHTTPRouteValidationWebhook(t *testing.T) {
...
+				Rule: admregv1.Rule{
+					APIGroups:   []string{"gateway.networking.k8s.io"},
+					APIVersions: []string{"v1beta1"},
+					Resources:   []string{"httproutes"},
+				},
+				Operations: []admregv1.OperationType{admregv1.Create, admregv1.Update},
+			},
+		},
+	)
+	assert.NoError(t, err, "creating webhook config")
```
Comment (pmalek) [inline test/integration/httproute_webhook_test.go:233]: Why no 

```suggestion
	require.NoError(t, err, "creating webhook config")
```

?

File: test/integration/httproute_webhook_test.go [hunk 5]
```diff
@@ -139,13 +88,197 @@ func TestHTTPRouteValidationWebhook(t *testing.T) {
...
+	t.Cleanup(func() {
+		assert.NoError(t, closer())
+	})
+
+	t.Log("creating a gateway client")
+	gatewayClient, err = gatewayclient.NewForConfig(env.Cluster().Config())
+	require.NoError(t, err)
+
+	ns, cleaner := helpers.Setup(ctx, t, env)
+	namespace = ns.Name
```
Comment (pmalek) [inline test/integration/httproute_webhook_test.go:243]: ditto, log the namespace name

File: test/integration/helpers_test.go [hunk 6]
```diff
@@ -420,14 +421,36 @@ func setIngressClassNameWithRetry(ctx context.Context, namespace string, ingress
...
 
+type routerFlavor string
+
+const (
+	traditional           routerFlavor = "traditional"
+	traditionalCompatible routerFlavor = "traditional_compatible"
+	expressions           routerFlavor = "expressions"
+)
+
+func skipTestForRouterFlavor(t *testing.T, flavor ...routerFlavor) {
```
Comment (pmalek) [inline test/integration/helpers_test.go:432]: This already deserves it's own file I believe ( together with other related `func`s ).

If it's not going to be used in other types of tests (other than integration tests) than it can be placed in this dir.

File: internal/validation/gateway/httproute.go [hunk 7]
```diff
@@ -181,3 +187,55 @@ func getListenersForHTTPRouteValidation(sectionName *gatewayv1beta1.SectionName,
 
 	return listenersForValidation, nil
 }
+
+func validateWithKongGateway(
+	ctx context.Context, routesSvc kong.AbstractRouteService, parserFeatures parser.FeatureFlags, kongVersion semver.Version, httproute *gatewayv1beta1.HTTPRoute,
```
Comment (pmalek) [inline internal/validation/gateway/httproute.go:192]: I'd advocate for trying to decouple this func signature from `kong.AbstractRouteService` and hence using an `interface` which we define here.

File: internal/validation/gateway/httproute.go [hunk 8]
```diff
@@ -181,3 +187,55 @@ func getListenersForHTTPRouteValidation(sectionName *gatewayv1beta1.SectionName,
...
+	if len(errMsgs) > 0 {
+		return false, validationMsg(errMsgs), nil
+	}
+	// Validate by using feature of Kong Gateway.
+	for _, kg := range kongRoutes {
+		kg := kg
+		ok, msg, err := routesSvc.Validate(ctx, &kg)
+		if err != nil {
+			return false, "unable to validate HTTPRoute schema", nil //nolint:nilerr
+		}
```
Comment (pmalek) [inline internal/validation/gateway/httproute.go:228]: - It might be beneficial to log the `err`. WDYT?
- What if we can a transient network error? Do we retry or fail validation hence rejecting the `HTTPRoute`?

File: internal/validation/gateway/httproute.go [hunk 9]
```diff
@@ -181,3 +187,55 @@ func getListenersForHTTPRouteValidation(sectionName *gatewayv1beta1.SectionName,
...
+			err    error
+		)
+		translation := translators.KongRouteTranslation{
+			Name:    "validation-attempt",
+			Matches: rule.Matches,
+			Filters: rule.Filters,
+		}
+		routes, err = parser.GenerateKongRouteFromTranslation(
+			httproute, translation, parserFeatures.RegexPathPrefix, parserFeatures.ExpressionRoutes, kongVersion,
+		)
```
Comment (pmalek) [inline internal/validation/gateway/httproute.go:210]: Why not just

```suggestion
		translation := translators.KongRouteTranslation{
			Name:    "validation-attempt",
			Matches: rule.Matches,
			Filters: rule.Filters,
		}
		routes, err := parser.GenerateKongRouteFromTranslation(
			httproute, translation, parserFeatures.RegexPathPrefix, parserFeatures.ExpressionRoutes, kongVersion,
		)
```

File: internal/validation/gateway/httproute.go [hunk 10]
```diff
@@ -39,11 +61,10 @@ func ValidateHTTPRoute(httproute *gatewayv1beta1.HTTPRoute, attachedGateways ...
 		}
 	}
 
-	// validate that no unsupported features are in use
-	if err := validateHTTPRouteFeatures(httproute); err != nil {
-		return false, "httproute spec did not pass validation", err
+	// Validate by sending converted routes to validation endpoint of Kong Gateway.
+	if routesSvc != nil {
+		return validateWithKongGateway(ctx, routesSvc, parserFeatures, kongVersion, httproute)
```
Comment (pmalek) [inline internal/validation/gateway/httproute.go:66]: I don't have an immediate solution for this but I'm not a huge fan of an implicit dependency like the `routesSvc` param 🤔 

Perhaps it would be better to create dedicated type(s) which could have a specific purpose of validating against Kong Gateway or not. 🤔

File: CHANGELOG.md [hunk 11]
```diff
@@ -78,6 +78,9 @@ Adding a new version? You'll need three changes:
 
 ### Added
 
+- Add validation in admission webhook for regex expressions supplied in HTTPRoute
+  (stop rejecting such configurations).
```
[the remark this answers is not in the document]
Comment (programmer04) [reply CHANGELOG.md:82]: We did both, I have to write it better. By the way, it's interesting that any user has hit this limitation before...

File: test/integration/helpers_test.go (hunk 6 above)
Comment (programmer04) [reply test/integration/helpers_test.go:432]: yeah I agree, I'm moving it separate file, btw. I would advocate for putting all of such helpers in one folder (not as a part of this PR), because now it's not obvious at first glance if the file contains tests or some helpers

File: internal/validation/gateway/httproute.go (hunk 8 above)
Comment (programmer04) [reply internal/validation/gateway/httproute.go:224]: In terms of retries etc. for such cases I rely on go-kong implementation it's done the same way in others validators too. Maybe it's worth to adress in separate PR. Btw. I see area for small refactors in this part of our codebase to do as standalone task.

Instead of logging I think we can return it

File: internal/validation/gateway/httproute.go (hunk 10 above)
Comment (programmer04) [reply internal/validation/gateway/httproute.go:66]: Yup, when I wrote this I thought to myself it's not great, but also not terrible, see what I'm proposing now, using small no-op validator

File: CHANGELOG.md [hunk 12]
```diff
@@ -78,6 +78,9 @@ Adding a new version? You'll need three changes:
 
 ### Added
 
+- Allow regex expressions in `HTTPRoute` configuration and provide validation in admission webhook.
+  Before this change admission webhook used to reject entirely such configurations incorrectly as not supported yet.
+  [#4608](https://github.com/Kong/kubernetes-ingress-controller/pull/4608)
```
Comment (pmalek) [inline CHANGELOG.md:83]: 👍

File: test/integration/webhook_test.go [hunk 13]
```diff
@@ -684,12 +680,10 @@ func waitForWebhookServiceConnective(ctx context.Context, configResourceName str
 	return networking.WaitForConnectionOnServicePort(waitCtx, env.Cluster().Client(), consts.ControllerNamespace, svcName, svcPort, 10*time.Second)
 }
 
-func ensureAdmissionRegistration(ctx context.Context, configResourceName string, rules []admregv1.RuleWithOperations) (func() error, error) {
+func ensureAdmissionRegistration(ctx context.Context, t *testing.T, configResourceName string, rules []admregv1.RuleWithOperations) {
 	svcName := fmt.Sprintf("webhook-%s", configResourceName)
 	svcCloser, err := ensureWebhookService(ctx, svcName)
```
Comment (pmalek) [inline test/integration/webhook_test.go:685]: We can make an action item or issue to refactor all those `ensure...()` calls to use `t.Cleanup`

File: test/integration/skip_helpers_test.go [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+	routerFlavor := routerFlavor(eventuallyGetKongRouterFlavor(t, proxyAdminURL))
+	for _, f := range flavor {
+		if routerFlavor == f {
+			t.Skipf("router flavor:%q skipping", f)
+		}
+	}
+}
+
+// Expression router is not supported for some objects and features.
+// For example, KongIngress is not supported by intention;
```
Comment (pmalek) [inline test/integration/skip_helpers_test.go:30]:
```suggestion
// For example, KongIngress is intentionally unsupported;
```

File: test/integration/skip_helpers_test.go [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+
+type routerFlavor string
+
+const (
+	traditional           routerFlavor = "traditional"
+	traditionalCompatible routerFlavor = "traditional_compatible"
+	expressions           routerFlavor = "expressions"
+)
+
+func skipTestForRouterFlavor(t *testing.T, flavor ...routerFlavor) {
```
Comment (pmalek) [inline test/integration/skip_helpers_test.go:19]: Since one can pass multiple values here I'd suggest to rename this

```suggestion
func skipTestForRouterFlavors(t *testing.T, flavor ...routerFlavor) {
```

File: internal/validation/gateway/httproute_test.go [hunk 16]
```diff
@@ -255,113 +257,6 @@ func TestValidateHTTPRoute(t *testing.T) {
 			validationMsg: "httproute spec did not pass validation",
 			err:           fmt.Errorf("queryparam matching is not yet supported for httproute"),
 		},
-		{
-			msg: "if an HTTPRoute is using regex path matching it fails validation due to lack of support",
```
Comment (pmalek) [inline internal/validation/gateway/httproute_test.go:259]: While removing this: do we have a test case here that would have a route with regex path  which passes the test? 🤔

Comment (pmalek) [reply internal/validation/gateway/httproute_test.go:259]: Oh is it now in `commonHTTPRouteValidationTestCases` ?

File: test/integration/webhook_test.go (hunk 13 above)
Comment (programmer04) [reply test/integration/webhook_test.go:685]: I'm doing PR for this (almost as fast as writing an issue), I will push it after merging this PR to avoid confilicts

File: internal/validation/gateway/httproute_test.go (hunk 16 above)
Comment (programmer04) [reply internal/validation/gateway/httproute_test.go:259]: That validation has been removed in this PR because it had been supported since closing
- https://github.com/Kong/kubernetes-ingress-controller/issues/2153
It's somewhat surprising that no user has noticed it.

Now it's allowed and covered with a test defined in `commonHTTPRouteValidationTestCases` 
https://github.com/Kong/kubernetes-ingress-controller/blob/39ab171288ca6bc994ed1f4fe4bddcf6746ce3c7/test/integration/httproute_webhook_test.go#L92-L117

File: test/integration/skip_helpers_test.go [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+	traditionalCompatible routerFlavor = "traditional_compatible"
+	expressions           routerFlavor = "expressions"
+)
+
+func skipTestForRouterFlavors(t *testing.T, flavor ...routerFlavor) {
+	t.Helper()
+	routerFlavor := routerFlavor(eventuallyGetKongRouterFlavor(t, proxyAdminURL))
+	for _, f := range flavor {
+		if routerFlavor == f {
+			t.Skipf("router flavor:%q skipping", f)
```
Comment (pmalek) [inline test/integration/skip_helpers_test.go:24]: Might as well print the URL here to help debugging

File: test/integration/httproute_webhook_test.go [hunk 18]
```diff
@@ -139,13 +87,199 @@ func TestHTTPRouteValidationWebhook(t *testing.T) {
...
+		},
+	)
+
+	t.Log("creating a gateway client")
+	gatewayClient, err := gatewayclient.NewForConfig(env.Cluster().Config())
+	require.NoError(t, err)
+
+	ns, cleaner := helpers.Setup(ctx, t, env)
+	namespace = ns.Name
+	t.Logf("created namespace: %q", namespace)
```
Comment (pmalek) [inline test/integration/httproute_webhook_test.go:240]: Sorry for that: it seems that https://github.com/Kong/kubernetes-ingress-controller/blob/d781042d455b82e079194728d72c4d451f9422d6/test/internal/helpers/setup.go#L41 already prints the namespace and the test name

File: test/integration/httproute_webhook_test.go [hunk 19]
```diff
@@ -139,13 +87,199 @@ func TestHTTPRouteValidationWebhook(t *testing.T) {
...
+				},
+			},
+			WantCreateErr:          true,
+			WantCreateErrSubstring: "regex parse error:\n    bar[[\n        ^\nerror: unclosed character class)",
+		},
+	)
+	testHTTPRouteValidationWebhook(ctx, t, namespace, gatewayClient, testCases)
+}
+
+func setUpEnvForTestingHTTPRouteValidationWebhook(ctx context.Context, t *testing.T) (
```
Comment (pmalek) [inline test/integration/httproute_webhook_test.go:212]: One more suggestion for improvement for this would be to add namespace and/or object selectors in `ensureAdmissionRegistration` so that all those webhook tests that run on the same cluster didn't interfere with each other and we get a bit more isolation 😓 

If that's going to be done in a separate PR then 👍 but I'd suggest we create an issue for that.

Comment (pmalek) [reply test/integration/httproute_webhook_test.go:212]: And also that would need to be mentioned in the godoc for this func

File: test/integration/httproute_webhook_test.go (hunk 18 above)
Comment (programmer04) [reply test/integration/httproute_webhook_test.go:240]: 👍 so reverting

File: test/integration/httproute_webhook_test.go (hunk 19 above)
Comment (programmer04) [reply test/integration/httproute_webhook_test.go:214]: https://github.com/Kong/kubernetes-ingress-controller/issues/4621

**Review by pmalek — APPROVED** [review]
